sqlglot/optimizer/annotate_types.py

```diff
@@ -568,48 +568,85 @@ def _annotate_literal(self, expression: exp.Literal) -> exp.Literal:
 
         return expression
 
+    def _maybe_wrap_with_array(self, expression: E, array: bool = False) -> E:
+        if array:
+            self._set_type(
+                expression,
+                exp.DataType(
+                    this=exp.DataType.Type.ARRAY,
+                    expressions=[expression.type],
+                    nested=True,
+                ),
+            )
+
+        return expression
+
     @t.no_type_check
     def _annotate_by_args(
         self,
         expression: E,
-        *args: str,
+        *args: str | exp.Expression,
         promote: bool = False,
         array: bool = False,
     ) -> E:
-        expressions: t.List[exp.Expression] = []
+        literal_type = None
+        non_literal_type = None
+
         for arg in args:
-            arg_expr = expression.args.get(arg)
-            expressions.extend(expr for expr in ensure_list(arg_expr) if expr)
+            if isinstance(arg, str):
+                expressions = expression.args.get(arg)
+            else:
+                expressions = arg
 
-        last_datatype = None
-        for expr in expressions:
-            expr_type = expr.type
+            for expr in ensure_list(expressions):
+                expr_type = expr.type
 
-            # Stop at the first nested data type found - we don't want to _maybe_coerce nested types
-            if expr_type.args.get("nested"):
-                last_datatype = expr_type
-                break
+                # Stop at the first nested data type
+                if expr_type.args.get("nested"):
+                    self._set_type(expression, expr_type)
+                    return self._maybe_wrap_with_array(expression, array=array)
```

Comment:
Should we just break here? That'd remove the need for the additional helper to wrap with an array.